### DISCUSSION TITLE
Proposal: Permit trailing commas in method signatures and invocations

### DISCUSSION THREAD

**Discussion by MgSam** [dotnet/csharplang#8638] [Language Ideas]
Migrated from https://github.com/dotnet/roslyn/issues/12496
### Proposal

Allow trailing commas in method/constructor declarations and invocations.

``` C#
void Foo(
    int bar, 
    int baz,
) { }

Foo(
    bar: 5,
    baz: 3,
);
```
### Rational

One very small but appreciated feature in C# is the tolerance of the compiler for trailing commas in array initialization or object initialization:

``` C#
var a = new int[] 
{ 
    3, 
    4, 
    5, 
};
var b = new 
{ 
    Bark = "woof", 
    Bite = "chomp", 
}
```

This makes it easier to reorder or edit the item order without taking time to make sure that last comma gets deleted. 

It would be nice if this feature also extended to method signatures and invocation. It would be especially useful in the post C# 4.0 where named parameters can be reordered or used to make a highly descriptive method call.
### Example

``` C#
void Foo (int bar, String baz, Func<int> cat, Action<String> dot, int elf) { }

Foo(
    bar: 5, 
    baz: "",
    cat: () => 6,
);
```

Wait, having `cat` last is confusing, let me place it after `bar`...

``` C#
//Rather than just copy and paste the relevant line...
Foo(
    bar: 5, 
    cat: () => 6, //Now I need to add this comma here!
    baz: "" //And I need to remove this comma here!
);
```

Instead, if method invocations were also tolerant of an extra comma, it would make editing easier:

``` C#
Foo(
    bar: 5, 
    cat: () => 6, 
    baz: "", //Extra comma here, it's not hurting anyone!
);
```
### Final thoughts
[TypeScript](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#trailing-commas-in-function-parameter-and-argument-lists) has and soon [JavaScript](http://2ality.com/2015/11/trailing-comma-parameters.html) will have this feature. I think it will work great in C# as well.

**Comment by DavidArno** [comment]
```cs
    baz: "", //Extra comma here, it's not hurting anyone!
```
It's hurting readability, simply to make the lives of lazy developers easier. It's bad enough that C# already supports superfluous commas in some cases already. Extending that to other cases would be very sad development.

**Comment by Pzixel** [comment]
@MgSam it's easy to write and hard to read. As you know, 80% of time you are reading the code. Thus this should be never get implemented.

**Comment by jveselka** [comment]
@DavidArno 
> It's hurting readability

How? Whats so confusing about it?

> ...simply to make the lives of lazy developers easier

Of course - thats what all language features are about

> It's bad enough that C# already supports superfluous commas in some cases already.

Its already allowed at least on enums and object initializers and how is it bad there? Did it ever caused you any issues?


Frankly, I do not understand why would anyone oppose this improvement.

**Comment by wanton7** [comment]
I think IDE should be responsible of adding and removing commas when you paste something. Example if you paste something inside list or array initialization. But this shouldn't be added to language.
TypeScript probably added this because it's in some future JavaScript spec.
This just adds noise to language (hurts readability).
This probably wouldn't be so big of an issue if C# had standard formatter like golang has. If you add this that comma will be all over the place.

I can see this happening

```
var a = new int[] 
{ 
    3, 
    4, 
    5
, 
};
```

**Comment by alrz** [comment]
Trailing commas are usually allowed when it doesn't affect type or binding. e.g. object initializers, probably `match`, etc. In an argument list it does affect binding (because of overload resolution), however, with named arguments it is not always significant.

**Comment by DavidArno** [comment]
@zippec,
> > It's hurting readability
> How? Whats so confusing about it?

I refer you to [my answer to a near identical question from a year ago](https://github.com/dotnet/roslyn/issues/8818#issuecomment-200329241):

-----
> "I'm aware it's optional; that's it's problem. When looking at a piece of code with a trailing comma, one must ask: is it there because the author likes to use them, because they forgot to refactor, or because something has been missed and thus it could be a bug. They create a serious hindrance to readability just to pander to developers who want their lives made fractionally easier when writing the code."
-----

> Frankly, I do not understand why would anyone oppose this improvement.

Because some of us see their use as indicative of sloppy programming practices that favour making code easy to write over easy to read.

**Comment by svick** [comment]
@alrz 

> In an argument list it does affect binding (because of overload resolution)

Do you have an example? I tried [a trivial example](https://tryroslyn.azurewebsites.net/#f:r/MYGwhgzhAEDC0G8CwAoa7oDcD2BLAJtALIAUuAdgC7RgCUiAvqhlnoaRdWADTSfQAjegiZoMOAsTAUSw5iwykALACZutANzz0ohkA===) of adding the trailing comma in current compiler and the error I got was:

> error CS1525: Invalid expression term ')'

It seems the compiler never reached the overload resolution phase.

**Comment by orthoxerox** [comment]
I have to write a lot of SQL at work, and the lack of trailing commas is the bane of every SQL programmer (any probably of every programmer that has to write JSON, too). While I do not think the arguments are edited *that* often to really require support for trailing commas, I also do not oppose this idea.

**Comment by alrz** [comment]
@svick

I meant the number of arguments, for example you could add property initializers without changing the return type, so it's safe to allow it there. But if you add an argument it might change the selected method.

**Comment by HaloFour** [comment]
punctuation means something to scatter it about just to placate formatting or decades-old diff technology is silly we have refactoring tools which manage the proper placement of commas on reordering members or arguments

**Comment by jnm2** [comment]
I'm torn. On one hand trailing commas irritate me whenever I see them; on the other hand, it would be more consistent with the rest of the language.

Much as I instinctively dislike this, punctuation exists to serve the wishes of those who use it and no one would be forcing me to use it that way.

**Comment by alrz** [comment]
As it is a matter of code style, an analyzer could preserve, insert, remove or prohibit usages in the code, so I wouldn't mind even if it would be allowed everywhere in the language that we have a list-like construct.

**Comment by MgSam** [comment]
@HaloFour It has nothing to do with diff tools. Anyone who does a lot of editing of code with long method invocations with many named parameters runs into this. As @orthoxerox said, the problem is even more acute in SQL, and in my personal experience, some of those invocations of which I speak are calls to SQL stored procs or tables.

The fact that JS is adding this feature in 2017 means that a committee comprised heavily from the biggest tech companies ratified it. That's a pretty big hurdle to cross. It's not some poorly thought out artifact left over from the language's inception like some other JS features. I think that's a pretty good indication that the use case here is real and pervasive.

**Comment by HaloFour** [comment]
@MgSam 

> Anyone who does a lot of editing of code with long method invocations with many named parameters runs into this.

Indeed I have.  Many, many, many times.  Particularly with SQL and particularly with stored procs with dozens of parameters.  Or tables with dozens to hundreds of fields.  It's of the most minor of inconveniences and not one that I feel deserves special punctuation rules.  Developers should probably be forced to manually fix their code more often after a bad copy&paste.

**Comment by MgSam** [comment]
It's definitely minor, and it only takes a second or two to correct. But correcting interrupts your train of thought. And when this happens a few times a day, everyday, the time adds up.

**Comment by jnm2** [comment]
@MgSam If the IDE magically handled it for you, that would solve your problem.

**Comment by alrz** [comment]
>magically 


Can you elaborate on that?

**Comment by wanton7** [comment]
@MgSam You talked about JSON before having same problem. Visual Studio 2017 already adds and removes commas when you paste to JSON array.  Visual Studio also adds quotes if paste plain text to an JSON array and probably more.

@alrz This same kind of mechanism could be added for C# arrays. It's magic ;)

**Comment by aluanhaddad** [comment]
While I think this will do little harm, it is a bad idea.

The argument I always hear in favor of this, time and time again, is that it affects multiple lines of code in a diff. I wont belabor this as @HaloFour already put it quite eloquently
> punctuation means something to scatter it about just to placate formatting or decades-old diff technology is silly we have refactoring tools which manage the proper placement of commas on reordering members or arguments

but considering the number of programming languages, let alone natural languages, that use commas to enumerate lists, it really is amazing that standard diff tools and formatters have not become more sophisticated. Lets not provide yet another reason for the evolution of our tools to stagnate.

@eyalsk 
> It should word like this: JavaScript gonna have this feature, therefor, TypeScript will. 😉

Truer words were never spoken!

While on the subject of JavaScript (TypeScript added this because JavaScript added it), if this is added, it will create yet another point of contention around code style and there are frankly enough of them already. As soon as this was added to JavaScript, a lot of people wasted a lot of time debating it.

**Comment by AlenPelin** [comment]
This feature must be added to the standard due to a number of reasons:

* C# already allows that in arrays and objects 
* it makes diff cleaner (that makes eyes happier)
* it removes certain merge conflicts (that require human attention and is a great source of bugs)
* other languages already / will have it

P.S. I'd also add a compiler switch that makes trailing commas essential everywhere.

**Comment by AlenPelin** [comment]
As for IDE handling that - in general case, we cannot rely on having IDE in all cases at all stages for everybody.

**Comment by Pzixel** [comment]
@AlenPelin it's looks like curing the headash by heading yourself... Traling comma looks like empty statement which is pretty valid and which is the source of great amount of bugs. Tool like R# underline them and suggest to remove...

I'm totally against it. If you add a new line to some dictionary in every commit then you probably doing something wrong. It happens, but it shouldn't enforce an error prone feature in the language.

**Comment by yaakov-h** [comment]
```csharp
void Foo (int bar, String baz, Func<int> cat) { }

var a = new int[] 
{ 
      3 
    , 4
    , 5
};

var b = new 
{ 
      Bark = "woof"
    , Bite = "chomp"
};

Foo(
      bar: 5
    , baz: ""
    , cat: () => 6
);
```

**Comment by AlenPelin** [comment]
```c#
void Foo (
    int bar, 
    string baz, 
    Func<int> cat, 
) { }

var a = new int[] 
{ 
    3, 
    4,
    5,
};

var b = new 
{ 
    Bark = "woof",
    Bite = "chomp",
};

Foo(
    bar: 5,
    baz: "",
    cat: () => 6,
);
```

**Comment by AlenPelin** [comment]
@Pzixel could you elaborate on what you meant?

> Traling comma looks like empty statement which is pretty valid and which is the source of great amount of bugs

Could you provide couple examples of bugs it can cause? 

> Tool like R# underline them and suggest to remove...

Only because it is not supported by language/compiler. For arrays and properties it does not underline them by default.

> I'm totally against it. If you add a new line to some dictionary in every commit then you probably doing something wrong.

Your position is understandable, and you are lucky that you can configure R# to treat trailing commas as errors to prevent compiling. Unlike you, me and others who want trailing commas in methods don't have such an option. In my opinion, it is not fair and needs to be fixed.

**Comment by HaloFour** [comment]
What I really don't like about this is that it's an example of modifying the language syntax to satisfy bad tooling, whether that be refactoring tools (or rather the lack of them) or diff tools.    The fact that the language does allow it today in limited circumstances is a source of unwanted inconsistency.  I'm not sure which I consider the worse sin.

**Comment by AlenPelin** [comment]
@HaloFour agree, tools must better support language and do not require that. In real world though tools are always imperfect and it is sad when you cannot use great tool just because it does not support certain C# limitations.

**Comment by Pzixel** [comment]
@AlenPelin you already have several options:

1. Start new line with commas just like with dots. No one cares about LINQ because most of queries are written in `dots first` style.
2. Append new lines at beginning instead of the end like i did yesterday:
![image](https://user-images.githubusercontent.com/11201122/31992849-1d002a3c-b984-11e7-8a2c-31553f8ff454.png)
3. Don't care.

I personally prefer the combination of 2nd and 3rd.

**Comment by alrz** [comment]
@HaloFour 

Tooling is a part of the ecosystem and actually an important scenario when it comes to language design, remember one of the challenges that delayed generators *is* tooling. I'd argue that the syntax should be developer-friendly at the end (not merely tool-friendly), the team is willing to change the language towards that goal, as they did for nullable reference types which is totally practical as a sole tooling feature just like R# demonstrates. But that wouldn't be much convinient.

I don't know how you imagine this could be addressed with "better" tooling. specialized diff algorithm? some editor features? those have much greater cost and wouldn't be universal. you should use a particular editor, doesn't seem to be an attractive solution (or rather, workaround).

**Comment by CyrusNajmabadi** [comment]
I would personally like this; primarily because of consistency:

1.  The consistency between all lists.  i.e. we already support this for arrays, why not for things like parameters?
2.  The consistency in code when writing things out.  If i have a list of things, one per line, it seems unfortunately inconsistent that one item is different than all the rest.

This also makes things more annoying when working with effectively *any* tool.  If i just want to swap lines around then I'm inconvenienced anytime i'm working with last and second-to-last elements.

**Comment by yaakov-h** [comment]
And ultimately anyone who doesn’t like trailing commas can write an analyzer. 😉

**Comment by CyrusNajmabadi** [comment]
Experimenting with the code change here: https://github.com/dotnet/roslyn/pull/22851/files

Looks like it would be fairly simple to implement.

**Comment by MgSam** [comment]
Nice. Now you just have to slip it into 7.3, Cyrus. :smiley:

**Comment by AlenPelin** [comment]
@Pzixel 

> Append new lines at beginning instead of the end like i did yesterday

This won't work at all with method declaration as the order matters:
```c#
void Welcome(
  string name, 
  string title
) { ... }

void Welcome(
  string title,
  string name
) { ... }
```
This won't work with method call unnamed arguments as the order matters:
```c#
void Welcome(string name, string title) { ... }

Welcome(
  "John Smith",
  "MR"
);

Welcome(
  "MR",
  "John Smith"
);
```

**Comment by orthoxerox** [comment]
Shouldn't they be allowed in tuples, too?

**Comment by AlenPelin** [comment]
@orthoxerox why not? But that's slightly out of this subject "Permit trailing commas in method signatures and invocations"

**Comment by DavidArno** [comment]
@yaakov-h,

We could make it a compiler switch... 😉

**Comment by yaakov-h** [comment]
@DavidArno `<Features>pedantic-developer-compat</Features>`

**Comment by Pzixel** [comment]
@AlenPelin good point. However, 99% of methods in my practice are singleliners because they don't have more than 4 arguments which is concidered a bad practice. I agree that sometimes these arguments are like`Expression<Func<IEnumerable<KeyValuePair<TKey, TValue>>>>>`, but it doesn't happen very often :)

However, I see that trailing commas allowing to make a better diff, but makes it harder to read. You make this diff once but read this line many many times after, and every time you see comma you want to see some continuation but there isn't. I condider it's harmful :)

**Comment by alrz** [comment]
If we want this for *all* syntax lists, I think we should include declarators as well,

```cs
const int 
     A = 1,
     B = 2,
     ;
```

**Comment by DavidArno** [comment]
Has anyone mentioned extends/implements lists yet?
```cs
class Foo : IFoo, Bar,
{
```
And let's not forget property patterns too:
```cs
if (x is Foo(X: var x, Y: var y,)) ...
```
In fact, whilst we are at it, why not just allow a trailing comma on every line of code, just in case?
😱

**Comment by jnm2** [comment]
@DavidArno I was just going to say, why not do it for generic arities as well? =P

`ValueTuple<IEnumerable<Foo>, Bar, IDictionary<string, Baz>, >`

**Comment by DavidArno** [comment]
@jnm2,

That's a good point. My "trailing comma on every line" idea would sadly be a breaking change here:
```cs
Type d1 = typeof(Dictionary<,,
>);
```
Oh well, back to just hating this idea. 😆

**Comment by aluanhaddad** [comment]
Maybe one day, diff tools will understand what lists are. Then they would have a mode where only one line is _shown_ as changed.

To put it another way, maybe they can evolve to have a notion of _logical_ lines instead of _physical_  lines, kind of like how hardware abstraction layers expose information to an operating system.

**Comment by CyrusNajmabadi** [comment]
> If we want this for all syntax lists, I think we should include declarators as well,

My impl would allow for this.

> Has anyone mentioned extends/implements lists yet?

My impl would allow for this.  

> I was just going to say, why not do it for generic arities as well?

This and tuples would worry me.  I can envision us doing things in the future where generic arities might be inferred, or tuple values might be omitted.  I can see having a trailing comma having meaning.  So i don't want to do this yet.

**Comment by CyrusNajmabadi** [comment]
> Maybe one day, diff tools

We have to accept that there is a broad set of tooling out there that has existed for *decades* and which has not done these "really nice to have things that would make the world come together and make a happier place".  Sometimes the path of least resistance and most effectiveness is just to accept that and make the change in the single tool itself.

This is similar to when i added support to C# to understand git-merge-markers.  Yes, we could hope all tools out there would understand and process them properly.  But that might take decades.  Or, we could just do the couple of hours of work and make the experience substantially better *today*.

I far prefer pragmatic solutions that bring about things now, versus waiting years.  

--

BTW: i've worked in this area for 15 years now.  I remember these same discussions 15 years ago.  And even back then people had been talking about them for ages.  It's just not going to happen for a long long time :)

**Comment by CyrusNajmabadi** [comment]
Another case i didn't want to change was commas in a lambda.  i.e allowing  ```(a, b, c, ) => ...```  I could imagine us allowing omitted parameters here in the future, and i don't want to make this ambiguous.

**Comment by jnm2** [comment]
> > I was just going to say, why not do it for generic arities as well?
> 
> This and tuples would worry me. 

Good, that was the motivation for me saying it. =)

**Comment by jveselka** [comment]
@DavidArno 

> "I'm aware it's optional; that's it's problem. When looking at a piece of code with a trailing comma, one must ask: is it there because the author likes to use them, because they forgot to refactor, or because something has been missed and thus it could be a bug. They create a serious hindrance to readability just to pander to developers who want their lives made fractionally easier when writing the code."

I can make the same argument about dozens of existing and completely fine constructs:
```C#
if (condition)
{
   action();
}
```
Are the unnecessary brackets "there because the author likes to use them, because they forgot to refactor, or because something has been missed and thus it could be a bug."

If you approach code with assumption that it is wrong, of course you end up with a lot of questions.

**Comment by ufcpp** [comment]
I often generate source codes like the following:

```cs
void Method(
    T1 arg1,
    T2 arg2,
    T3 arg3,
    T4 arg4,
    bool dummy = false)
```

I really want this proposal.

**Comment by DavidArno** [comment]
@zippec,

If I saw:
```cs
if (condition)
    action1();
    action2();
```
then I'd see a very likely bug: the author meant for `action2()` to only run when `condition` is true, which isn't the case. But your example is just "meh" to me; there is no confusion and it certainly isn't equivalent to the hideousness of trailing commas. 

Plus "the language is already full of mistakes; what harm some more?" is a pretty weak argument.

**Comment by JeroMiya** [comment]
I don't really experience the issue with diffs, but I definitely noticed this issue when attempting to refactor some code to use read-only data structures instead of get/set properties. I would previously use the object initializer:

```CSharp
var foo = new Foo() {
  Property1 = "value1",
  Property2 = "Value2",
};
```
And this was nice, especially early in the development of a feature as I add or remove or re-order properties from the list, I don't have to waste time and attention maintaining the comma on the "last property" in the list here. It's just part of my muscle memory to add a comma to every property in the list. That was the whole point of C# being permissive with that last comma in object and array and dictionary initializers. It's a great feature of C# that is a good quality of life improvement.

But now, when I refactor the above code to use read-only data structures, say for a uni-directional data flow pattern (so I can enforce read-only state at the compiler level, instead of merely relying on convention), then I lose that "last line comma" permissiveness, and just from personal experience I can tell you I REALLY notice the difference and it's extremely annoying:

```CSharp
// aside: C# *really* needs a better syntax for read-only records like this, like F# has.
// read-only data structures in C# require *painful* amounts of boilerplate.
// so much so that I often seriously consider defining them all in an F# class library.
class Foo() {
  public Foo(string property1 = null, string property2 = null, string property3 = null)
  {
    Property1 = property1;
    Property2 = property2;
    Property3 = property3;
  }
  public string Property1 { get; }
  public string Property2 { get; }
  public string Property3 { get; }
}
var foo = new Foo(
  property1: "value1",
  property2: "value2"
);
```

To be honest, having to maintain that last line in a comma separated list of function arguments is one of my pet peeves, just as it is with object and array initializers, and is just one of several syntactic speed bumps for creating read-only data structures in C#.

Oh, and this:
```CSharp
var foo = new Foo(
    property1: "value1"
  , property2: "value2"
);
```
Is not a tangible improvement. You've just moved the comma maintenance to the first line instead of the last line.

Is this is a high priority feature? No.
Would it make me happier? Yes.
Would it make me measurably more productive? Marginally, yes.

Does it result in ambiguities?
Not if you ignore the last comma in an argument list, when no value is provided:

```
// semantically, these two statements are equivalent, and should result in the same AST:
Foo(1, 2);
Foo(1, 2,);
```

I am also OK with people using them in argument lists on the same line, especially if there are more than three arguments:
```
// Now I can add/remove/reorder this argument list with a simple copy/paste
FunctionWithLotsOfArguments(arg1: 1, arg2: 2, arg3: 3, arg4: 4,); 
```

**Comment by MgSam** [comment]
Kotlin just added this feature as well.

https://kotlinlang.org/docs/reference/whatsnew14.html#trailing-comma

**Comment by CyrusNajmabadi** [comment]
I'm championing this.

**Comment by Pzixel** [comment]
It's funny to see that 3 years later after making an acquaintance with several languages including Rust, Idris and some others I've changed my position in 180 degrees and now I think it's pretty valuable, for better git diffs at least 😃

**Comment by alrz** [comment]
@CyrusNajmabadi Would this be only about method arguments/parameters or all the lists? e.g. tuples, type parameter constraints, and so on. (for an existing instance I think Rust does permit all those)

**Comment by YairHalberstadt** [comment]
This is particularly useful for source generators, where having to deal with the lack of a trailing comma tends to add a few extra lines of code to any loop printing parameter or argument lists. Whilst on each case its simple, overall it adds up to a lot of clutter in the source generator.

**Comment by datvm** [comment]
I want to add a use case for this suggestion, for the new `record` syntax, this would be awesome (currently impossible):

```
    public record Foo (
        int bar1,
        string? bar2,
    );
```

**Comment by mikernet** [comment]
There is another case where I think a trailing comma should be allowed that I don't believe was mentioned: collection initializer `Add` method parameters, especially when they are `params`. For example:

```cs
public class Simple : IEnumerable<(string, string, string)>
{
    public void Add(string x, string y, string z) { }
}

var a = new Simple() {
    {
        "fkeuahwkeufhawefe",
        "afkuehafkuehkuehf",
        "aswefa4rguikareg", // this comma is not allowed
    },
};

public class WithParams : IEnumerable<IReadOnlyList<string>>
{
    public void Add(params string[] x) { }
}

var b = new WithParams () {
    {
        "fkeuahwkeufhawefe",
        "afkuehafkuehkuehf",
        "aswefa4rguikareg", // this comma is not allowed
    },
    {
        "fkeuahwkeufhawefe",
        "afkuehafkuehkuehf", // this comma is not allowed
    },
    {
        "fkeuahwkeufhawefe",
        "afkuehafkuehkuehf",
        "fkeuahwkeufhawefe",
        "afkuehafkuehkuehf",
        "fkeuahwkeufhawefe",
        "afkuehafkuehkuehf",
        "aswefa4rguikareg", // this comma is not allowed
    },
};
```

**Comment by Pzixel** [comment]
I think trailing comma should be available in *any* context that accepts array of arguments. It could be just an optional trivia for any `SeparatedList`

**Comment by MixusMinimax** [comment]
I strongly agree with this. For the people that find that an extra comma hurts readability, how are you able to read the arguments that come before? Those have a comma after them, and you don't complain there.

Merge conflicts are a huge issue. Sure, there are syntax-aware diff systems, which I hope become more prevalent in the future, but for now, diffs are line based. Adding a new field to a record should not modify _two_ lines.

**Comment by theunrepentantgeek** [comment]
> For the people that find that an extra comma hurts readability, how are you able to read the arguments that come before? Those have a comma after them, and you don't complain there.

My objection comes from the cognitive science around how people read.

For people whose first reading language uses a comma as a phrase separator (*), the presence of a `,` is a subconscious visual indicator that there's more to come in this sentence.

(*) From what I've read, it's less clear whether this applies to people whose first language has different forms.

When you read `foo, bar, baz` each comma is understood by your brain as indicating there's more to come.

When you read `foo, bar, baz,` your brain has to do a doubletake because the promise made by the `,` that there's more to come is immediately violated.

Redundant trailing commas are a cognitive speedbump. There's very strong evidence from the natural language field that shows incorrect punctuation greatly slows down comprehension. 

My personal experience working in Go backs this up. It's like having an uneven brick in your garden path that you trip on most every time you check the letterbox.

> Merge conflicts are a huge issue. Sure, there are syntax-aware diff systems, which I hope become more prevalent in the future, but for now, diffs are line based. 

And this is exactly why we **should not** be changing the language.

Language changes are forever; they can never be undone. 

Why should we make a permanent change to the language just to accommodate tooling that hasn't yet caught up?

Language changes are supremely expensive, both to make and to maintain. It would be cheaper to fix the tooling.

**Comment by jnm2** [comment]
My experience is that ite roadbump disappears quickly with use. I switched over for multiline initializers, where the comma appears consistently at the end of the lines which end each item.

**Comment by Socolin** [comment]
For me it would remove the "noise" of all the removed/added lines due to the addtion of a coma, during code review.

Example:

```diff
public SomeClass(
	IDeps1 deps1,
-	IDeps2 deps2
+	IDeps2 deps2,
+	IDeps3 deps3,
)
```

I would need this only for multi line initializer, it does not make sence in single line. And like `;` at the end of each line, I think the brain can understand that `,` at the end of each line of a parameters is a part of the syntax.

```cs
public SomeClass(
	IDeps1 deps1,
	IDeps2 deps2,
	IDeps3 deps3,
)
```

@theunrepentantgeek I'm not sure to understand what the point with one-liner example, since this feature really seems to be useful in a multiline example (I don't see the point in adding extra `,` if the code is on one line), and I may be wrong but when I'm seeing this, I see this as a list and not as a sentense for which I' m looking for "what next"

```
foo,
bar,
baz, 
```



> Language changes are supremely expensive, both to make and to maintain. It would be cheaper to fix the tooling.

Changing all the diff tool and merge tool and git blame, to undestand that specific case to improve the diff does not seems cheaper. And I don' t think git will start parsing C# and understant that a coma added between 2 parameters should be ignored anytime soon.

**Comment by Logerfo** [comment]
Reducing noise in diffs is the main reason I always use trailing commas in multi-line scenarios. Allowing them in parameter and argument lists would cover this noticeable feature gap.

**Comment by JeroMiya** [comment]
I don't think I would want a diff tool to hide a changed line, even if it were capable of it through syntax aware diffing.

**Comment by theunrepentantgeek** [comment]
> I'm not sure to understand what the point with one-liner example

I was using a single line because I was giving a natural language example, not a code one.

That said, C# has never been a whitespace sensitive language, so it's irrelevant whether we're talking single- or multi-line. If a trailing comma is permitted in one place, it will be permitted in the other.

> Changing all the diff tool and merge tool and git blame, to undestand that specific case to improve the diff does not seems cheaper. 

Language changes are extremely expensive, which is why they get discussed and debated endlessly before they're made.

My guestimate is that the cost of changing the language would exceed the cost of updating the tools by at least an order of magnitude (that is, 10x).

> My experience is that ite roadbump disappears quickly with use.

We developers can get used to most anything. We're smart like that. 

In this case, however, given the hestitation that is built into the way our brains parse language (at an autonomic level), it's not something we can ever avoid, but instead just something we get used to. 

I suspect this kind of adaptation is like learning to avoid the squeaky step on a staircase: you stop irritating people in the middle of the night, but the problem is still there.

I've gotten used to trailing commas in Go, simply because I have no choice. But that doesn't mean I like them - because I'm aware of the science that makes them a foolish quirk of the language.

**Comment by Socolin** [comment]
While speaking about tooling, if the only reason not to add the coma is a preference and readability issue, I think then we should allow this and add a setting in the IDEs to show/hide the trailing comma. 

This would allow to fix problem with the diff tools and not add new problem for humans.

The comma could be added automatically and hidden depending on the settings of the project / IDE

**Comment by miloush** [comment]
I wouldn't support IDE hiding parts of content of the file, but aren't all these cognitive arguments resolved by just not using the feature if it makes the experience worse for you?

**Comment by snebjorn** [comment]
JavaScript was able to do [trailing commas](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas) and it's awesome! JavaScript pretty must runs the world. so if that was able to do it then C# also should 😄 

I see comments about "preference and readability". Lets not forget that the trailing commas are **optional**. If you don't like them or it makes it harder for you to read. Then don't use them.

**Comment by MixusMinimax** [comment]
Cognitive science is mainly to be applied to natural languages. A list of arguments is not a sentence, it is a list of arguments. And yes, if you don't like it, then don't write trailing commas, that's what formatters are there for to enforce. But the arguments for merge conflicts, consistency, etc. strongly outweigh "idk it's not as nice to read". That's an argument to not _use_ the feature, sure.

**Comment by jnm2** [comment]
> That said, C# has never been a whitespace sensitive language, so it's irrelevant whether we're talking single- or multi-line.

For questions of formatting and readability, whitespace has never been irrelevant.

**Comment by theunrepentantgeek** [comment]
> Lets not forget that the trailing commas are optional. If you don't like them or it makes it harder for you to read. Then don't use them

This is only relevant for toy projects that will only ever have a single developer. 

As soon as you start working on anything with a meaningful size, you're looking at contributions from multiple developers over multiple years. 

Later devs don't get a choice, they have to deal with whatever their predecessors have done. 

Arguments of "If you don't like it, don't use it" always founder on this fact.

**Comment by snebjorn** [comment]
> As soon as you start working on anything with a meaningful size, you're looking at contributions from multiple developers over multiple years.
> 
> Later devs don't get a choice, they have to deal with whatever their predecessors have done.

Code style is a team decision. They can change in the lifetime of a project. When they do you run the formatting tool with the new settings. I don't see a problem here.

**Comment by mrwensveen** [comment]
For me, the inconsistency between initializing a record type (constructor) and a class with properties annoys me the most because the use case is so similar. If it only were to be allowed in collection initializers then I would be okay with it (kind of).

Some people have argued that it's bad enough that C# allows it in other places but that's a matter of opinion. The feature clearly made it through, presumably with people arguing for and against the feature.

So, while I think this feature would be nice, I feel more strongly about the language being inconsistent with regards to trailing comma's.

**Comment by bw-flagship** [comment]
The (quite new) language "dart" allows trailing commas. IDEs make use of this and automatically break lines if you add a trailing comma on method calls, constructors, etc. Switching from back to coding c# from coding dart feels like a downgrade because this is not possible.

**Reply by pharan** [reply]
Dart feels so much like a second attempt at a next-generation Java (like C# tried to do originally).
C# could learn a lot from Dart syntax-wise. Trailing commas is an excellent example of something that fundamentally improves the ease of making changes to code safely there and I do miss it and final declarations when switching back to C#.

**Comment by KennethHoff** [comment]
With the introduction of Primary constructors I think this "issue" should be looked at. You may wonder what Primary Constructors have to do with anything, so here's my argument (pun intended):

Previously whenever you wanted to add or remove a field from a class you had to do it many places; Add the field, add a parameter to the constructor and set the field from the parameter (And obviously the same thing in reverse for removing), so adding/removing a single comma wasn't that big of a deal.

Now on the other hand you only have to change a single line to do all those three things mentioned previously, so the removal of a comma is now a majority of the time spent. Instead of simply pressing `shift-delete` (default keybind in Rider for deleting an entire line), you also have to go up a line and remove the comma.

(This is in addition to all the other arguments mentioned above, like consistency with initializers)

**Comment by beppemarazzi** [comment]
Or, in other words, in many context (i.e. initialization list, argument list, ...) the comma will act as a "terminator" not as a "separator". Exactly like the semicolon does with statements. (https://en.wikipedia.org/wiki/Comparison_of_programming_languages_(syntax)#:~:text=A%20statement%20separator%20demarcates%20the,end%20of%20an%20individual%20statement.).
It's another (little) point in favor of this proposal to make the whole language more self consistent...

**Comment by Bigsby** [comment]
This would improve code versioning because, when adding items to lists (e.g., function parameters) only the added line is marked as changed and not the previous last line where the comma was added.

**Comment by TheBrambleShark** [comment]
Coming over from #8190

This is very useful when targeting multiple TFMs. Take this example that's currently impossible today:
```cs
/// <summary>
/// Represents a 9-digit United States social security number.
/// </summary>
[PublicAPI]
[DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(),nq}}")]
public readonly struct SocialSecurityNumber
    : IFormattable,
    IEquatable<SocialSecurityNumber>,
#if NET7_0_OR_GREATER
    IEqualityOperators<SocialSecurityNumber, SocialSecurityNumber, bool>,
#endif
#if NET8_0_OR_GREATER
    IUtf8SpanFormattable,
    IUtf8SpanParsable<SocialSecurityNumber>
#endif
{
    // Specification omitted
}
```
Such is only possible with trailing commas allowed.

There are a couple work-arounds, such as SQL-style commas, but style rules may prohibit this:
```cs
/// <summary>
/// Represents a 9-digit United States social security number.
/// </summary>
[PublicAPI]
[DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(),nq}}")]
public readonly struct SocialSecurityNumber
#pragma warning disable SA1001 // Commas should not be preceeded by whitespace.
    : IFormattable
    , IEquatable<SocialSecurityNumber>
#if NET7_0_OR_GREATER
    , IEqualityOperators<SocialSecurityNumber, SocialSecurityNumber, bool>
#endif
#if NET8_0_OR_GREATER
    , IUtf8SpanFormattable
    , IUtf8SpanParsable<SocialSecurityNumber>
#endif
```
Partial classes are another option, but this doesn't make sense in all cases.

**Comment by mikernet** [comment]
@Foxtrek64 Just FYI, you can also do:

```cs
public readonly struct SocialSecurityNumber :
#if NET7_0_OR_GREATER
    IEqualityOperators<SocialSecurityNumber, SocialSecurityNumber, bool>,
#endif
#if NET8_0_OR_GREATER
    IUtf8SpanFormattable,
    IUtf8SpanParsable<SocialSecurityNumber>,
#endif
    IFormattable,
    IEquatable<SocialSecurityNumber>
{
}
```

**Comment by aradalvand** [comment]
The use cases are patently clear and justified. Source generators, TFMs, consistency, diffing, etc. This one little trivial-to-implement feature, which almost every other modern language already has, would improve the experience in all those areas.

So why not just do it and get it over with?

**Comment by CyrusNajmabadi** [comment]
> So why not just do it and get it over with?

Because our plates are full with lots of other work.

**Comment by miyu** [comment]
If this is implemented, it'd be consistent to support trailing commas in array indexers too. This would be useful for a variety of niche DSLs that I've seen... Trivial examples would be DSLs which index by filter condition or slice N-dimensional data grids.
